lib/SPIRV/SPIRVReader.cpp

```diff
@@ -685,49 +689,145 @@ void SPIRVToLLVM::setLLVMLoopMetadata(const LoopInstType *LM, Instruction *BI) {
                           LoopControlParameters[++NumParam])));
         break;
       }
-      default:
```

Comment:
>   if (LC & LoopControlPeelCountMask) {
    ++NumParam;
    assert(NumParam <= LoopControlParameters.size() &&
           "Missing loop control parameter!");
  }
  if (LC & LoopControlPartialCountMask && !(LC & LoopControlDontUnrollMask)) {
    // If unroll factor is set as '1' - disable loop unrolling
    if (1 == LoopControlParameters[NumParam])
      Metadata.push_back(getMetadataFromName("llvm.loop.unroll.disable"));
    else
      Metadata.push_back(llvm::MDNode::get(
          *Context,
          getMetadataFromNameAndParameter("llvm.loop.unroll.count",
                                          LoopControlParameters[NumParam])));
    ++NumParam;
    assert(NumParam <= LoopControlParameters.size() &&
           "Missing loop control parameter!");
  }

If NumParam == 4 and LoopControlParameters.size() == 4 then assert on line 657 will be passed. But then we may read from memory which is out of LoopControlParameters bound.

Reply:
The block where this memory is read is not entered unless the correspondent field of the loop control mask is set. This comment does raise an important point, however: I may want to add an assertion somewhere after line 695 to guard against undefined behavior on an invalid SPIR-V.